API update

BasileiosKal opened the pull request:
Updating the way that the messages and generators are passed to the operations, using @andrewwhitehead's proposal of passing them in pairs. 

The main purpose of the PR is to have a concrete base of discussion about what the best API will be. For alternatives see #117. For some more context see PR #128 (the motivation was “how we get the revealed messages from the (msg_1, …, msg_L) list and the revealed indexes).

IMO all approaches have trade-offs. Although I like the proposal seen in this PR a lot and it makes things more explicit, I wonder if it is more error prone?? That said, it may provide the best balance between being explicit and easy to read. Interested on the WG feedback.

andrewwhitehead commented on draft-bbs-signatures.md:
```diff
@@ -389,15 +389,14 @@ Procedure:
 - header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.
-- msg_1,..., msg_L (OPTIONAL), a vector of octet strings.
-- H_1,..., H_L (OPTIONAL), points of G1. Generators used to commit each message.
+- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list with pairs of the form (point in G1, scalar). Generators and messages pairs, corresponding to messages to be signed. If not supplied, it will default to the empty array ("[]"). The list MUST be ordered based on the values of the generators  (i.e., H_1 < H_2 < ... < H_L).
```
The list MUST be ordered based on the index of the generators?

BasileiosKal replied:
This does make sense considering how the generators are created but that order is harder to be "verified" unless the verifier re-creates all the generators (not really a security issue I think, it's just that the proof will not validate). 

Used the suggestion for now and opened issue #141 to discuss more.

andrewwhitehead commented on draft-bbs-signatures.md:
```diff
@@ -498,22 +496,28 @@ Procedure:
+                                        corresponding to messages to be revealed. 
+                                        Generators must be G1 points, and the 
+                                        messages must be non-zero scalars mod q.
+- [(H_h1, msg_h1), ..., (H_hU, msg_hU)] (REQUIRED), Generators and messages pairs 
```
Not sure why these two are required and can't default to [] as well

BasileiosKal replied:
My thinking was that if both lists are optional, it would allow both to be empty at the same time (which it should not if the original signed list is also non-empty). 

That said, I added a not in the description of ProofGen to highlight this fact and made the lists optional. I agree that this way is better from having the lists be required.

andrewwhitehead commented on draft-bbs-signatures.md:
```diff
@@ -524,7 +528,8 @@ Parameters:
 
 - L, is the non-negative integer representing the number of messages to be signed e.g length(msg_1,...,msg_L). Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).
-- R, is the non-negative integer representing the number of revealed messages e.g length(RevealedIndexes). Note if no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
+- R, is the non-negative integer representing the number of revealed messages e.g length(RevealedIndexes). Note if no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).H
```
RevealedIndexes doesn't exist anymore

BasileiosKal replied:
Thanks! Addressed.

andrewwhitehead commented:
Thanks for that, it looks pretty readable to me although maybe we can find a better way to describe the ordering of the generators.

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -534,75 +569,79 @@ Procedure:
+     messages must be non-zero scalars mod q. If not supplied, it will default to the 
+     empty array ("[]"). The list MUST be ordered based on the index of the 
+     generators.
+- [H_h1, ..., H_hU] (OPTIONAL), a list of points in G1. Generators corresponding to 
+     messages to be kept hidden. If not supplied, the list will default to the empty 
+     array ("[]"). The list MUST be ordered based on the index of the generators.
```
```suggestion
- [H_h1, ..., H_hU] (OPTIONAL), a list of points in G1. Generators corresponding to 
     un-revealed messages. If not supplied, the list will default to the empty 
     array ("[]"). The list MUST be ordered based on the index of the generators.
```

I think reveal vs un-reveal is better language than adding in another term "hidden"?

BasileiosKal replied:
Thanks! Addressed

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -612,8 +651,20 @@ Parameters:
+- R, is the non-negative integer representing the number of revealed messages e.g.,
+     length([(H_i1, msg_i1), ..., (H_iR, msg_iR)]). If no revealed messages are
+     supplied as an input to this operation, the value of R MUST evaluate to zero (0).
+- U, is the non-negative integer representing the number of non-revealed messages e.g.,
```
```suggestion
- U, is the non-negative integer representing the number of un-revealed messages e.g.,
```

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -389,15 +389,20 @@ Procedure:
+     information. If not supplied, it defaults to an empty string.
+- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generators and message pairs 
+     corresponding to messages to be signed. Generators must be G1 points, and the 
+     messages must be non-zero scalars mod q. If not supplied, it will default to the 
```
will need to be mod r if #140 is merged first

tplooker replied:
will need to be mod r if #140 is merged first

tplooker commented on draft-bbs-signatures.md:
````diff
@@ -498,22 +511,32 @@ Procedure:
 ```
-proof = ProofGen(PK, signature, header, ph, (msg_1,..., msg_L), (H_1,..., H_L), RevealedIndexes)
+proof = ProofGen(PK, signature, header, ph,
+                 [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
````
[(H_1, msg_1, reveal?) ... (H_L, msg_L, reveal?)] an alternative syntax suggestion

tplooker replied:
[(H_1, msg_1, reveal?) ... (H_L, msg_L, reveal?)] an alternative syntax suggestion

tplooker commented on draft-bbs-signatures.md:
````diff
@@ -534,75 +569,79 @@ Procedure:
 
 ```
-result = ProofVerify(PK, proof, ph, header, (msg_i1,..., msg_iR), RevealedIndexes, (H_1,..., H_L))
+result = ProofVerify(PK, proof, header, ph,          
+                 [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
+                 [H_h1, ..., H_hU])
````
```suggestion
result = ProofVerify(PK, proof, header, ph,          
                 [H_1, ..., H_L], 
                 [(index, msg_x), ..., (index, msg_x)])
```

Alternative syntax suggestion
```suggestion
result = ProofVerify(PK, proof, header, ph,          
                 [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
                 [H_h1, ..., H_hU])
```

tplooker replied:
```suggestion
result = ProofVerify(PK, proof, header, ph,          
                 [H_1, ..., H_L], 
                 [(index, msg_x), ..., (index, msg_x)])
```

Alternative syntax suggestion
```suggestion
result = ProofVerify(PK, proof, header, ph,          
                 [(H_i1, msg_i1), ..., (H_iR, msg_iR)], 
                 [H_h1, ..., H_hU])
```

BasileiosKal commented:
Updated API based on the discussion on the  WG call of 16/5/2022.

The main change is the proposal of the `[(H_i, msg_i), for revealed_i = 1]` etc notation. It may not be 100% formally defined but IMO it is “clear enough” while avoiding some of the issues we discussed on the call. 

My biggest concern here is that the order of the revealed messages supplied to ProofVerify is not clearly enough defined. But we 
may be able to solve this with some additional text?

Interested on the WG’s thoughts. c.c. @tplooker, @andrewwhitehead ^

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -498,22 +496,21 @@ Procedure:
-- msg_1,..., msg_L (OPTIONAL), octet strings. Messages in input to Sign.
-- H_1,..., H_L (OPTIONAL), points of G1. The generators in input to Sign.
-- RevealedIndexes (OPTIONAL), vector of unsigned integers. Indexes of revealed messages.
+- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting by a generator, the message and a variable indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ...
```
```suggestion
- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting of a generator, the message and a boolean value indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```
```suggestion
- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting by a generator, the message and a variable indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```

tplooker replied:
The last statement here around the ordering I still dont think is 100% clear, but I think in general this PR is a great improvement.

tplooker replied:
The last statement here around the ordering I still dont think is 100% clear, but I think in general this PR is a great improvement.

BasileiosKal replied:
Changed the statement a bit (now the ordering needs to follow the ordering of the input to Sign) and provided some examples to make it more clear.

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -498,22 +496,21 @@ Procedure:
-- msg_1,..., msg_L (OPTIONAL), octet strings. Messages in input to Sign.
-- H_1,..., H_L (OPTIONAL), points of G1. The generators in input to Sign.
-- RevealedIndexes (OPTIONAL), vector of unsigned integers. Indexes of revealed messages.
+- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting by a generator, the message and a variable indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ...
```
```suggestion
- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting of a generator, the message and a boolean value indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```
```suggestion
- [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples consisting by a generator, the message and a variable indicating if the message will be revealed or not. Generators must be G1 points, messages must be non-zero scalars mod r and the variable indicating if that message will be revealed must be boolean. If not supplied, the list will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```

BasileiosKal replied:
Done! Thanks!

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -534,9 +532,9 @@ Procedure:
 1. signature_result = octets_to_signature(signature)
 
-2. (i1, i2,..., iR) = RevealedIndexes
+2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_i, msg_i), if revealed_i = 1]
```
I think this syntax needs to be explained more, in general we could elect to do more of a filter or map style function?

tplooker replied:
```suggestion
2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_i, msg_i), if revealed_i = 1]
```
```suggestion
2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)].filter(item => item.revealed_i == true)
```

tplooker replied:
```suggestion
2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_i, msg_i), if revealed_i = 1]
```
```suggestion
2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)].filter(item => item.revealed_i == true)
```

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -534,9 +532,9 @@ Procedure:
 1. signature_result = octets_to_signature(signature)
 
-2. (i1, i2,..., iR) = RevealedIndexes
+2. [(H_i1, msg_i1), ..., (H_iR, msg_iR)] = [(H_i, msg_i), if revealed_i = 1]
```
I think this syntax needs to be explained more, in general we could elect to do more of a filter or map style function?

BasileiosKal replied:
I think that the complexity of having to define a filter function will outweigh the advantages of the current API??? In my last push I've described in "clear text" the operation of getting `(H_i1, msg_i1), ..., (H_iR, msg_iR)` etc., Thoughts on the alternative???

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -523,20 +520,27 @@ Parameters:
-- L, is the non-negative integer representing the number of messages to be signed e.g length(msg_1,...,msg_L). Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).
-- R, is the non-negative integer representing the number of revealed messages e.g length(RevealedIndexes). Note if no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
+- L, is the non-negative integer representing the number of signed messages e.g., length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if no messages are signed, the value of L MUST evaluate to zero (0).
+- R, is the non-negative integer representing the number of revealed messages e.g., R = revealed_1 + revealed_2 + ... + revealed_L. If no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
```
```suggestion
- R, is the non-negative integer representing the number of revealed messages e.g., R = revealed_1 + revealed_2 + ... + revealed_L. In the calculation of R its assumed that the boolean value of true evaluates to 1 whereas the boolean value of false evaluates to 0.  If no messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
```

tplooker replied:
Trying to clarify that effectively we are casting boolean values in the calculation of R to an integer?

BasileiosKal replied:
Good point! Updated to a different wording that it is sorter but may not be as clear?? If yes i will update to the suggestion.

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -523,20 +520,27 @@ Parameters:
 
-- L, is the non-negative integer representing the number of messages to be signed e.g length(msg_1,...,msg_L). Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).
-- R, is the non-negative integer representing the number of revealed messages e.g length(RevealedIndexes). Note if no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
+- L, is the non-negative integer representing the number of signed messages e.g., length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if no messages are signed, the value of L MUST evaluate to zero (0).
```
```suggestion
- L, is the non-negative integer representing the number of signed messages, length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if no messages are signed, the value of L MUST evaluate to zero (0).
```

tplooker replied:
Its not an example IMO so we should drop the e.g

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -523,20 +520,27 @@ Parameters:
-- R, is the non-negative integer representing the number of revealed messages e.g length(RevealedIndexes). Note if no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
+- L, is the non-negative integer representing the number of signed messages e.g., length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if no messages are signed, the value of L MUST evaluate to zero (0).
+- R, is the non-negative integer representing the number of revealed messages e.g., R = revealed_1 + revealed_2 + ... + revealed_L. If no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
+- U, is the non-negative integer representing the number of un-revealed messages e.g., U = L-R. If no non-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).
```
```suggestion
- U, is the non-negative integer representing the number of un-revealed messages, U = L-R. If no non-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).
```

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -387,15 +387,14 @@ Procedure:
 - header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.
-- msg_1,..., msg_L (OPTIONAL), a vector of octet strings.
-- H_1,..., H_L (OPTIONAL), points of G1. Generators used to commit each message.
+- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generators and message pairs corresponding to messages to be signed. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, it will default to the empty array ("[]"). The list MUST be ordered based on the index of the generators.
```
```suggestion
- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generators and message pairs corresponding to messages to be signed. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, it will default to the empty array ("[]").
```

Do we need the last sentence here?

BasileiosKal replied:
Indeed we don't. Thanks! Updated.

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -405,7 +404,7 @@ Parameters:
 Definitions:
 
-- L, is the non-negative integer representing the number of messages to be signed e.g length(msg_1,...,msg_L).  Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).
+- L, is the non-negative integer representing the number of messages to be signed i.e., length([(H_1, msg_1),..., (H_L, msg_L)]).  Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).
```
```suggestion
- L, is the non-negative integer representing the number of messages to be signed i.e., length([(H_1, msg_1),..., (H_L, msg_L)]).  Note if the list of generators and message pairs are not supplied as an input to this operation, the value of L MUST evaluate to zero (0).
```

BasileiosKal replied:
Done.

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -439,15 +438,14 @@ Procedure:
 - header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.
-- msg_1,..., msg_L (OPTIONAL), an optional vector of octet strings.
-- H_1,..., H_L (OPTIONAL), points of G1. Generators used to commit each message.
+- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generators and message pairs corresponding to signed messages. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, it will default to the empty array ("[]").
```
```suggestion
- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generator and message pairs corresponding to signed messages. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, it will default to the empty array ("[]").
```

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -457,7 +455,7 @@ Parameters:
 Definitions:
 
-- L, is the non-negative integer representing the number of messages to be signed e.g length(msg_1,...,msg_L). Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).
+- L, is the non-negative integer representing the number of messages to be signed i.e., length([(H_1, msg_1),..., (H_L, msg_L)]). Note if no messages are supplied as an input to this operation, the value of L MUST evaluate to zero (0).
```
```suggestion
- L, is the non-negative integer representing the number of messages to be signed i.e., length([(H_1, msg_1),..., (H_L, msg_L)]). Note if the list of generator and message pairs are not supplied as an input to this operation, the value of L MUST evaluate to zero (0).
```

BasileiosKal commented:
Couple of updates:
1. Updates based on the reviews.
2. Since the description for the list of (generator, message) pairs in the Input section for the ProofGen and ProofVerify operations was getting longer and longer, i changed the description to something more "minimal" and added a "see above for requirements" to point to the description above the each operation, where the input lists are explained clearly along with an example.

mikelodder7 commented on draft-bbs-signatures.md:
```diff
@@ -233,7 +233,7 @@ In definition of this signature scheme there are two possible variations based u
 ### Messages and generators
 
-Throughout the operations of this signature scheme, each message that is signed is paired with a specific generator (point in G1). Specifically, if a generator `H_1` is multiplied with `msg_1` during signing, then `H_1` MUST be multiplied with `msg_1` in all other operations as well (signature verification, proof generation and proof verification). For simplicity, each function will take as input ...
+Throughout the operations of this signature scheme, each message that is signed is paired with a specific generator (point in G1). Specifically, if a generator `H_1` is multiplied with `msg_1` during signing, then `H_1` MUST be multiplied with `msg_1` in all other operations as well (signature verification, proof generation and proof verification).
```
```suggestion
Throughout the operations of this signature scheme, each message that is signed is paired with a specific generator (point in G1). Specifically, if a generator `H_1` is multiplied with `msg_1` during signing, then `H_1` MUST be multiplied with `msg_1` in all other operations (signature verification, proof generation and proof verification).
```

BasileiosKal replied:
Done! Thanks!

mikelodder7 commented on draft-bbs-signatures.md:
```diff
@@ -515,20 +514,29 @@ Parameters:
-- R, is the non-negative integer representing the number of revealed messages e.g length(RevealedIndexes). Note if no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
+- L, is the non-negative integer representing the number of signed messages i.e., length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if the list of (generator, message boolean) tuples is not supplied, the value of L MUST evaluate to zero (0).
+- R, is the non-negative integer representing the number of revealed messages i.e., R is the number of the boolean values: revealed_1, ..., revealed_L, that are "true". If no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
+- U, is the non-negative integer representing the number of un-revealed messages i.e., U = L-R. If no non-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).
```
```suggestion
- U, is the non-negative integer representing the number of un-revealed messages i.e., U = L-R. If no un-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).
```

BasileiosKal replied:
Thanks! Updated!

BasileiosKal commented on draft-bbs-signatures.md:
```diff
@@ -439,15 +438,14 @@ Procedure:
 - header (OPTIONAL), an octet string containing context and application specific information. If not supplied, it defaults to an empty string.
-- msg_1,..., msg_L (OPTIONAL), an optional vector of octet strings.
-- H_1,..., H_L (OPTIONAL), points of G1. Generators used to commit each message.
+- [(H_1, msg_1),..., (H_L, msg_L)] (OPTIONAL), a list of generator and message pairs. Generators must be G1 points, and the messages must be non-zero scalars mod r. If not supplied, the list will default to the empty array ("[]"). The list MUST have the same order as when supplied to Sign.
```
There is non! other than it is not "technically required", after the adoption of the signature domain.

That said, there may be use-cases where people want to use the un-linkable presentations part of bbs and not the selective showing one (like OAuth bearer tokens for example??)

andrewwhitehead commented:
I think ProofGen turns out simpler if we don't try to combine the revealed message flags with the generators and message values. It means introducing more syntax plus the concept of boolean variables. For comparison...

> - [(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)] (OPTIONAL), a list of tuples containing a generator (point on G1), a non-zero scalar mod r representing a message, and a boolean value indicating if the message will be revealed or not (see above for requirements)
>
> ...
>
> - L, is the non-negative integer representing the number of signed messages i.e., length([(H_1, msg_1, revealed_1), ..., (H_L, msg_L, revealed_L)]). Note if the list of (generator, message boolean) tuples is not supplied, the value of L MUST evaluate to zero (0).
> - R, is the non-negative integer representing the number of revealed messages i.e., R is the number of the boolean values: revealed_1, ..., revealed_L, that are "true". If no revealed messages are supplied as an input to this operation, the value of R MUST evaluate to zero (0).
> - U, is the non-negative integer representing the number of un-revealed messages i.e., U = L-R. If no un-revealed messages are supplied as an input to this operation (i.e., if R = L), the value of U MUST evaluate to zero (0).
>
> ...
>
> 1. Let (H_i1, msg_i1), ..., (H_iR, msg_iR), be the pairs containing the
>    first 2 elements from the tuples (H_i, msg_i, revealed_i),
>    i = 1, ..., L, for which it holds that revealed_i = true, in the same
>    order (i.e., the (generator, message) pairs corresponding to the
>    revealed messages).
> 2. Let (H_j1, msg_j1), ..., (H_jU, msg_jU), be the pairs containing the
>    first 2 elements from the tuples (H_i, msg_i, revealed_i),
>    i = 1, ..., L, for which it holds that revealed_i = false, (i.e., the
>    (generator, message) pairs corresponding to the
>    un-revealed messages).

- [(H_1, msg_1), ..., (H_L, msg_L)] (OPTIONAL), a list of tuples containing a generator (point on G1), a non-zero scalar mod r representing a message.
- RevealedIndexes (OPTIONAL), an ordered list of distinct message indexes to be revealed. Each member of this list must be an integer such that 1 <= index <= L.

- L, is the non-negative integer representing the number of signed messages i.e., length([(H_1, msg_1), ..., (H_L, msg_L)]). In the absence of a list of (generator, message) tuples, this value is zero (0).
- R, is the non-negative integer representing the number of revealed messages i.e., length(RevealedIndexes). In the absence of a list of revealed message indexes, this value is zero (0).
- U, is the non-negative integer representing the number of un-revealed messages i.e., U = L-R.

andrewwhitehead commented:
As a follow-up to that suggestion, here it is with the variable alias:

proof = ProofGen(PK, signature, header, ph, signed=[], revealed=[])

Inputs:
...
- signed (OPTIONAL), a list of 2-tuples containing a generator (point on G1) and a message scalar (non-zero integer mod r), taking the form [(H_1, msg_1), ..., (H_L, msg_L)]. When unspecified, defaults to an empty list ([]).
- revealed (OPTIONAL), an ordered list of distinct message indexes to be revealed along with the proof. Each member of this list must be an integer such that 1 <= index <= L. When unspecified, defaults to an empty list ([]).

- L, the non-negative integer representing the number of signed messages, equal to length(signed).
- R, a non-negative integer representing the number of revealed messages, equal to length(revealed).
- U, a non-negative integer representing the number of un-revealed messages, equal to L-R.

BasileiosKal commented:
PR updated based on @andrewwhitehead's above proposal^ and ready for re-review. See #159 for considerations and comparisons between the two choices (boolean variables vs indexes).

andrewwhitehead commented on draft-bbs-signatures.md:
```diff
@@ -439,205 +452,269 @@ Procedure:
+3. [j1, ..., jU] = (1, ..., L) \ revealedIndexes
 
-4. if signature_result is INVALID, return INVALID
+4. (H_i1, msg_i1), ..., (H_iR, msg_iR) = signed[i1], ..., signed[iR]
```
Currently unused, but I guess it will be when the revealed messages and indexes are hashed

andrewwhitehead commented on draft-bbs-signatures.md:
```diff
@@ -439,205 +452,269 @@ Procedure:
-This operation checks that a proof is valid for a header, vector of revealed messages (along side their index corresponding to their original position when signed) and presentation header against a public key (PK).
+This operation checks that a proof is valid for a header, vector of generators, revealed messages, and presentation header against a public key (PK).
+
+The generators are supplied as a list (called `generators`) of points in G1. The generators MUST be the same as the ones supplied to the [Sign](#sign) operation, as part of the (generator, message) pairs of the `to_sign` input list. The generators list MUST also be sorted based on the order that the (generator, message) pairs had in that list. More specifically, if the list `to_sign = [(H_1, msg_ ...
```
*ProofVerify

andrewwhitehead commented on draft-bbs-signatures.md:
```diff
@@ -439,205 +452,269 @@ Procedure:
+
+The generators are supplied as a list (called `generators`) of points in G1. The generators MUST be the same as the ones supplied to the [Sign](#sign) operation, as part of the (generator, message) pairs of the `to_sign` input list. The generators list MUST also be sorted based on the order that the (generator, message) pairs had in that list. More specifically, if the list `to_sign = [(H_1, msg_ ...
+
+The revealed messages are supplied as a list of at most `length(generators)` pairs, of the form (revealed_message, index), containing a non-zero scalar mod r, representing a (revealed) message, and an integer in the 1 to L range, representing the index that the (generator, message) pair containing the revealed_message had in the `to_sign` list. The list MUST be sorted based on the (ascending) ord ...
```
It seems a bit more natural to use (index, message) to me given that in many languages that makes the list naturally sortable, sorting by the first and then the second element

andrewwhitehead commented on draft-bbs-signatures.md:
```diff
@@ -439,205 +452,269 @@ Procedure:
+7. (A', Abar, D, c, e^, r2^, r3^, s^, (m^_j1,...,m^_jU)) = proof
 
-5. generators =  (H_s || H_d || H_1 || ... || H_L)
+8. generators =  (H_s || H_d || H_1 || ... || H_L)
```
Shadows the previous `generators` variable

andrewwhitehead commented on draft-bbs-signatures.md:
```diff
@@ -439,205 +452,269 @@ Procedure:
+
+5. H_j1, ..., H_jU = generators[j1], ..., generators[jU]
+
+6. if KeyValidate(PK) is INVALID, return INVALID
```
This should probably remain near the beginning?

andrewwhitehead commented:
Thanks for the update! In general I do find this more understandable, but I'd like to hear what others think.

BasileiosKal commented:
Thanks @andrewwhitehead for the reviews. Updated based on your comments. 

Couple of additional changes:
1. added some more guidance on Sign and Verify on how the generators and the messages are to be supplied.
2. added an additional check on ProofVerify for the "correctness" of the indexes.

tplooker commented on draft-bbs-signatures.md:
```diff
@@ -386,50 +386,66 @@ Procedure:
-- Ciphersuite_ID (REQUIRED), ASCII string. The unique ID of the ciphersuite.
-- H_s (REQUIRED), point of G1. The generator for the blinding value of the signature.
-- H_d (REQUIRED), point of G1. The generator used to sign the signature domain.
+- Ciphersuite_ID (REQUIRED), ASCII string. The unique ciphersuite ID.
```
Can remove the notion of (REQUIRED) here as all parameters are required

BasileiosKal replied:
I'm happy to make the change if people thing that to not be that big of an issue.

BasileiosKal commented:
#### NOTE:
Waiting a little on resolution on #161 (or a consensus) to update this PR.

BasileiosKal commented:
Closing since updates are addressed by PR #183
